Marlin/src/feature/repeat.cpp

```diff
@@ -0,0 +1,62 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (c) 2020 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * Based on Sprinter and grbl.
+ * Copyright (c) 2011 Camiel Gubbels / Erik van der Zalm
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+#include "../inc/MarlinConfig.h"
+
+#if ENABLED(GCODE_REPEAT_MARKERS)
+
+#include "repeat.h"
+
+#include "../gcode/gcode.h"
+#include "../sd/cardreader.h"
+
+repeat_marker_t Repeat::marker[MAX_REPEAT_NESTING];
+int8_t Repeat::count, Repeat::index;
+
+bool Repeat::early_parse_M41(char * const cmd, const uint32_t &sdpos) {
+  parser.parse(cmd);
+  if (parser.command_letter == 'M' && parser.codenum == 41) {
+    gcode.process_parsed_command();
+    marker[count].sdpos = sdpos;
+    return false;
+  }
+  return true;
+}
+
+void Repeat::add_marker(const uint16_t count) {
+  if (count < MAX_REPEAT_NESTING) {
+    marker[count].counter = count;
+    count++;
+  }
+}
+
+void Repeat::loop() {
+  if (!count)                           // No marker has been set?
+    card.setIndex(0);                   // Go back to the top
+  else {
+    const uint16_t prev = count - 1;    // The marker
+    card.setIndex(marker[prev].sdpos);  // Change the SD card position now!
+    if (marker[prev].counter && !--marker[prev].counter) // If counter is running and reached zero...
+      --count;                          // ...The next `M41` will go to the previous marker, or 0
```

Comment:
I think there is a flow problem in this code.

I think that if you have nested loops, you won't be able to exit an inner loop and execute additional commands in the outer loop.

The last M41 in a loop probably needs to skip resetting the card index, to allow exiting a loop without forcing a jump to a previous marker (or the top of the file).